### PR TITLE
Spark: Adopt the new Scan Task APIs in Spark Readers

### PR DESCRIPTION
**Pull request by flyrain**
This PR mainly adopts the change in #5077.  I'm also experimenting the Change log reader to demonstrate how the new reader hierarchy works. 
cc @aokolnychyi @rdblue @RussellSpitzer @stevenzwu @szehon-ho @karuppayya

### CODE REVIEW

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/ChangelogRowReader.java
Comment (flyrain) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/ChangelogRowReader.java:35]: This class is only for demonstrating, I will remove it before merging, and will file a separated PR for change log reader.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java
Comment (flyrain) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java:100]: I'm not happy with this down casting. To eliminate it, we can have a method returning all related content files in class `ContentScanTask`. For example,
1. The method returns data file as well as all delete files in `FileScanTask`.
2. The method returns data file,  `addedDeletes` files, `existingDeletes` files and  in `DeletedRowsScanTask`

We name the method as `allContentFiles()`, `relatedContentFiles()`, etc

**Comment by aokolnychyi** [comment]
Let me take a look now.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java
Comment (szehon-ho) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java:92]: I guess it's already not fully functional style as we are re-assigning stream variable, so maybe we can simplify somewhat by just returning like:
if (contentScanTask.isFileScanTask()) {
   return Stream.concat...
} else if (contentScantask instanceof AddedRowsScanTask) {
   return stream.concat...

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java
Comment (szehon-ho) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java:104]: A little outside scope of change but now we are making a variable, how about, 

Map<String, ByteBuffer> keyMetadata = dataFileStream.collect.toMap(file -> file.key, file -> file.value)

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java
Comment (szehon-ho) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java:47]: Same suggestion for CST

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/RowDataRewriter.java
Comment (szehon-ho) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/RowDataRewriter.java:115]: Why is it necessary?

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/RowDataReader.java
Comment (szehon-ho) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/RowDataReader.java:46]: Is RowDataReader only for FileScanTask?  We should just provide the parameter to BaseDataReader right?

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java
Comment (szehon-ho) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java:124]: Is BatchDataReader only for FileScanTask?  Should we just pass this as parameter to BaseDataReader?

**Review by szehon-ho — COMMENTED** [review]
Took a preliminary look and made some comments

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java
Comment (aokolnychyi) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java:67]: To be honest, I am not convinced we have to restrict this to `ContentScanTask`. Also, keep in mind that different changelog tasks can be packed into the same task group. That's why the reader should not be restricted to strictly one task type.

I'd consider using `ScanTask` and changing the hierarchy a bit. This can can become `BaseReader`. I also don't think we need the second type parameter. We can just work with `ScanTaskGroup<ST>`.

```
abstract class BaseReader<T, ST extends ScanTask> implements Closeable {

  ...

  BaseReader(Table table, ScanTaskGroup<ST> taskGroup) {
    this.table = table;
    this.tasks = taskGroup.tasks().iterator();
    this.inputFiles = inputFiles(taskGroup);
    this.currentIterator = CloseableIterator.empty();
  }

  ...
}
```

Then I'd consider adding `BaseRowReader` like this (also no data in the name).

```
abstract class BaseRowReader<ST extends ScanTask> extends BaseReader<InternalRow, ST> {

  private final Schema tableSchema;
  private final Schema expectedSchema;
  private final String nameMapping;
  private final boolean caseSensitive;

  BaseRowReader(Table table, ScanTaskGroup<ST> taskGroup, Schema expectedSchema, boolean caseSensitive) {
    super(table, taskGroup);
    this.tableSchema = table.schema();
    this.expectedSchema = expectedSchema;
    this.nameMapping = table.properties().get(TableProperties.DEFAULT_NAME_MAPPING);
    this.caseSensitive = caseSensitive;
  }

  protected Schema tableSchema() {
    return tableSchema;
  }

  protected Schema expectedSchema() {
    return expectedSchema;
  }

  protected CloseableIterable<InternalRow> newIterable(InputFile file, FileFormat format, long start, long length,
                                                       Expression residual, Schema projection,
                                                       Map<Integer, ?> idToConstant) {
    switch (format) {
      case PARQUET:
        return newParquetIterable(file, start, length, residual, projection, idToConstant);

      case AVRO:
        return newAvroIterable(file, start, length, projection, idToConstant);

      case ORC:
        return newOrcIterable(file, start, length, residual, projection, idToConstant);

      default:
        throw new UnsupportedOperationException("Cannot read unknown format: " + format);
    }
  }

  ...
}
```

Then the existing `RowDataReader` won't need to change a lot.

```
class RowDataReader extends BaseRowReader<FileScanTask> {
  ...
}
```

Finally, we will have `ChangelogRowReader` capable of reading all types of changelog tasks like this.

```
class ChangelogRowReader extends BaseRowReader<ChangelogScanTask> {
  ...
}
```

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java
Comment (aokolnychyi) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java:100]: Yeah, this is not good. I'd be up exposing something like `referencedDataFiles` or `dataFiles` with `referencedDeleteFiles` or `deleteFiles` to avoid `?` in public API.

Maybe, we can even even add those methods to `ScanTask`?

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java
Comment (aokolnychyi) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java:172]: If we change the boundary to `ScanTask`, then this method can accept `ContentScanTask<?>`.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/ChangelogRowReader.java
Comment (aokolnychyi) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/ChangelogRowReader.java:37]: We can't have a reader per changelog scan task as a single task group can contain multiple task types.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/RowDataReader.java
Comment (flyrain) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/RowDataReader.java:46]: No. RowDataReader will also be inherited by cdc readers.

File: core/src/main/java/org/apache/iceberg/util/PartitionUtil.java
Comment (flyrain) [reply core/src/main/java/org/apache/iceberg/util/PartitionUtil.java:69]: @aokolnychyi and @szehon-ho are right. We only need to change the type. Will make the change.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java
Comment (flyrain) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java:67]: I agree that the second type parameter isn't needed. Thanks for the suggestion.

If we want to use `ScanTask` in the `BaseReader/BaseDataReader`, we need to either 
1. move some common logic to its subclasses, e.g. `getInputFile()`, `constantsMap()` , which need the method from `ContentScanTask`.
2. add new methods in `ScanTask`. e.g.  `file()`

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/RowDataRewriter.java
Comment (flyrain) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/RowDataRewriter.java]: Removed after the new change.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/ChangelogRowReader.java
Comment (flyrain) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/ChangelogRowReader.java:39]: This class is still experimental.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseReader.java
Comment (aokolnychyi) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseReader.java:80]: I think we should be able to get rid of changes in `ScanTask` and offer an abstract method like this:

```
protected abstract Stream<ContentFile<?>> referencedFiles(TaskT task);
```

We shouldn't be calling abstract methods in the constructor so you can make `inputFiles` lazy.

```
private final ScanTaskGroup<TaskT> taskGroup;
private final Iterator<TaskT> tasks;

...

private Map<String, InputFile> lazyInputFiles;

...

protected InputFile getInputFile(String location) {
  return inputFiles().get(location);
}

private Map<String, InputFile> inputFiles() {
  if (lazyInputFiles == null) {
    Stream<EncryptedInputFile> encryptedFiles = taskGroup.tasks().stream()
        .flatMap(this::referencedFiles)
        .map(file -> EncryptedFiles.encryptedInput(table.io().newInputFile(file.path().toString()), file.keyMetadata()));

    // decrypt with the batch call to avoid multiple RPCs to a key server, if possible
    Iterable<InputFile> decryptedFiles = table.encryption().decrypt(encryptedFiles::iterator);

    Map<String, InputFile> files = Maps.newHashMapWithExpectedSize(taskGroup.tasks().size());
    decryptedFiles.forEach(decrypted -> files.putIfAbsent(decrypted.location(), decrypted));
    this.lazyInputFiles = ImmutableMap.copyOf(files);
  }

  return lazyInputFiles;
}
```

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseReader.java
Comment (aokolnychyi) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseReader.java:114]: Are we sure we want to change `!currentTask.isDataTask()`?

On a side note, this can be rewritten with the new abstract method:

```
if (currentTask != null && !currentTask.isDataTask()) {
  String filePaths = referencedFiles(currentTask)
      .map(file -> file.path().toString())
      .collect(Collectors.joining(", "));
  LOG.error("Error reading file(s): {}", filePaths, e);
}
```

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseRowReader.java
Comment (aokolnychyi) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseRowReader.java:44]: I'd say either get rid of this var and use `table().schema()` or move it to `BaseReader` where we have other getters.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseRowReader.java
Comment (aokolnychyi) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseRowReader.java:74]: I know the old code was formatted this way but can we fix it now to match the method above?

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseRowReader.java
Comment (aokolnychyi) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseRowReader.java:92]: Same here

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseRowReader.java
Comment (aokolnychyi) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseRowReader.java:114]: Same here

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/RowDataReader.java
Comment (aokolnychyi) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/RowDataReader.java:59]: The old name seems a little bit weird. Shall we rename it to `inputFile`?

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/RowDataReader.java
Comment (aokolnychyi) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/RowDataReader.java:60]: I think the error message should be adapted now as the task is not necessarily FileScanTask

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/RowDataReader.java
Comment (aokolnychyi) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/RowDataReader.java:61]: What about using two separate return statements and pulling the format into a separate var like this?

```
if (task.isDataTask()) {
  return newDataIterable(task.asDataTask(), readSchema);
} else {
  InputFile inputFile = getInputFile(task.file().path().toString());
  Preconditions.checkNotNull(inputFile, "Could not find associated InputFile with task");
  FileFormat format = task.file().format();
  return newIterable(inputFile, format, task.start(), task.length(), task.residual(), readSchema, idToConstant);
}
```

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkDeleteFilter.java
Comment (aokolnychyi) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkDeleteFilter.java:32]: Do we still need this as a separate class?

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/ChangelogRowReader.java
Comment (aokolnychyi) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/ChangelogRowReader.java:39]: Let's add in a separate PR.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java
Comment (aokolnychyi) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java:44]: nit: does it have to be public?

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseRowReader.java
Comment (flyrain) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseRowReader.java:43]: not necessary, will make the change.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/ChangelogRowReader.java
Comment (flyrain) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/ChangelogRowReader.java:39]: Will remove it in this PR.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseReader.java
Comment (flyrain) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseReader.java:72]: Nice suggestion!

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/RowDataReader.java
Comment (flyrain) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/RowDataReader.java:60]: It is still FileScanTask here. 
```
class RowDataReader extends BaseRowReader<FileScanTask>
```

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkDeleteFilter.java
Comment (flyrain) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkDeleteFilter.java:32]: Moved it to `BaseReader`

**Review by aokolnychyi — COMMENTED** [review]
This looks almost ready to go. I had a question about constructing an empty delete filter for batch reads instead of passing null and a few nits.

File: spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/source/TestBaseReader.java
Comment (flyrain) [reply spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/source/TestBaseReader.java:56]: I'm not sure, but I prefer this renaming since we changed the class name from `BaseDataReader` to `BaseReader`.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseReader.java
Comment (flyrain) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseReader.java:242]: Make sense. I was trying not to touch `EqualityDeleteRowReader`. Made the change.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseReader.java
Comment (flyrain) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseReader.java:186]: No. Let me remove it.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java
Comment (flyrain) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java]: Agreed that it is safer to not change the behavior.

**Review by aokolnychyi — APPROVED** [review]
LGTM pending tests.

**Comment by aokolnychyi** [comment]
Thanks for the change, @flyrain! Thanks for reviewing, @szehon-ho!

**Pull request merged**
